Tests/NIOConcurrencyHelpersTests/NIOConcurrencyHelpersTests.swift

```diff
@@ -750,4 +756,273 @@ class NIOConcurrencyHelpersTests: XCTestCase {
+    }
+
+    @available(*, deprecated, message: "AtomicBox is deprecated, this is a test for the deprecated functionality")
+    func testLoadAndStoreHammering() {
+        let allDeallocations = NIOAtomic<Int>.makeAtomic(value: 0)
+        let iterations = 100_000
+
+        @inline(never)
+        func doIt() {
+            let box = AtomicBox(value: IntHolderWithDeallocationTracking(0, allDeallocations: allDeallocations))
+
+            spawnAndJoinRacingThreads(count: 6) { i in
+                switch i {
+                case 0: // writer
+                    for i in 1 ... iterations {
+                        box.store(IntHolderWithDeallocationTracking(i, allDeallocations: allDeallocations))
+                    }
+                default: // readers
+                    while true {
+                        if box.load().value < 0 || box.load().value > iterations {
+                            XCTFail("loaded the wrong value")
+                        }
+                        if box.load().value == iterations {
+                            break
+                        }
+                    }
+                }
+            }
+        }
+
+        doIt()
+        assert(allDeallocations.load() == iterations + 1, within: .seconds(1))
+    }
+
+    @available(*, deprecated, message: "AtomicBox is deprecated, this is a test for the deprecated functionality")
+    func testLoadAndCASHammering() {
+        let allDeallocations = NIOAtomic<Int>.makeAtomic(value: 0)
+        let iterations = 100_000
+
+        @inline(never)
+        func doIt() {
+            let box = AtomicBox(value: IntHolderWithDeallocationTracking(0, allDeallocations: allDeallocations))
+
+            spawnAndJoinRacingThreads(count: 6) { i in
+                switch i {
+                case 0: // writer
+                    for i in 1 ... iterations {
+                        let old = box.load()
+                        XCTAssertEqual(i - 1, old.value)
+                        if !box.compareAndExchange(expected: old,
+                                                   desired: .init(i, allDeallocations: allDeallocations)) {
+                            XCTFail("compare and exchange didn't work but it should have")
+                        }
+                    }
+                default: // readers
+                    while true {
+                        if box.load().value < 0 || box.load().value > iterations {
+                            XCTFail("loaded wrong value")
+                        }
+                        if box.load().value == iterations {
+                            break
+                        }
+                    }
+                }
+            }
+        }
+
+        doIt()
+        assert(allDeallocations.load() == iterations + 1, within: .seconds(1))
+    }
+
+    @available(*, deprecated, message: "AtomicBox is deprecated, this is a test for the deprecated functionality")
+    func testMultipleLoadsRacingWhilstStoresAreGoingOn() {
```

Comment:
Nice test, but could we make the loading loops use a while loop instead and ensure they actually keep reading until the very end? Currently they could be done much quicker than the writing thread.

Reply:
good point, done.